core/src/main/scala/org/apache/spark/status/AppStatusListener.scala

```diff
@@ -1208,6 +1232,33 @@ private[spark] class AppStatusListener(
     }
   }
 
+  private def killedTaskSummaryForSpeculationStageSummary(
+      reason: TaskEndReason,
+      oldSummary: Map[String, Int],
+      isSpeculative: Boolean): Map[String, Int] = {
+    reason match {
+      case k: TaskKilled if k.reason.contains("another attempt succeeded") =>
+        if (isSpeculative) {
+          oldSummary.updated("original attempt succeeded",
+            oldSummary.getOrElse("original attempt succeeded", 0) + 1)
+        } else {
+          oldSummary.updated("speculated attempt succeeded",
+            oldSummary.getOrElse("speculated attempt succeeded", 0) + 1)
+        }
+      // If the stage is finished and speculative tasks get killed, then the
+      // kill reason is "stage finished"
+      case k: TaskKilled if k.reason.contains("Stage finished") =>
+        if (isSpeculative) {
+          oldSummary.updated("original attempt succeeded",
+            oldSummary.getOrElse("original attempt succeeded", 0) + 1)
+        } else {
+          oldSummary
```

Comment:
If we reach this `else` block, is it correct to return `oldSummary`?
If the reason of `Stage finished` is that the speculative task finishes, should we do like [here](https://github.com/apache/spark/pull/33253/files#diff-0d63c25e93eb1fbec25754f0bee7714c7e3c1538bc7370d94b147524db5c1f63R1245-R1246)?

Reply:
So with `stage finished` reason if the task killed is speculative, then only we should increment `original attempt succeeded` not sure if the task which gets killed with `stage finished` is not speculative then we can safely increment the count for `speculative attempt succeeded` or not. WDYT?

Reply:
> not sure if the task which gets killed with stage finished is not speculative then we can safely increment the count for speculative attempt succeeded or not. WDYT?

Hmm, it seems difficult to judge that one attempt succeeded because the other fails (Both attempts can fail).

Reply:
> Hmm, it seems difficult to judge that one attempt succeeded because the other fails (Both attempts can fail).

Take back my comment. `Stage finished` is the reason when a task killed but the stage which the killed task belongs to successfully finished.

https://github.com/apache/spark/blob/2fe12a75206d4dbef6d7678b876c16876136cdd0/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L1641-L1675

Reply:
@sarutak missed to respond here, even if it is zombie `StageData` tracks metrics per stage attempt so the killed task could be either the original task or speculated task. So we are good right? I don't think we need special handling for zombie tasks. Am I missing something here?

Reply:
Sorry for the late reply. 
> I don't think we need special handling for zombie tasks.

Yes, I don't think we need to handle zombie tasks too. What I mean is that, if the reason is `Stage finished`, the killed task can be one of the following.

(1) Speculative attempt of a task
(2) A task which has speculative attempt
(3) A task which neither is speculative nor has speculative, but is zombie

For (1), it is handled [here](https://github.com/apache/spark/pull/33253/files#diff-0d63c25e93eb1fbec25754f0bee7714c7e3c1538bc7370d94b147524db5c1f63R1251).
I don't think we need to handle the case (3) but should we handle (2) in the [else block](https://github.com/apache/spark/pull/33253/files#diff-0d63c25e93eb1fbec25754f0bee7714c7e3c1538bc7370d94b147524db5c1f63R1255) somehow?

Reply:
@sarutak Do you mean handling the case where the task that got killed has a speculative attempt but got killed because of the speculative attempt succeeded? I think this probably would be difficult to address as the stage is finished and all the in memory data structures would be cleaned up, only the `AppStatusStore` would have the information. I don't think it is worth the effort to handle that checking `AppStatusStore`. In that case, the `speculation attempt succeeded` count can be one less, it is not a huge discrepancy IMO. But in the case of zombie task attempts the counters can get completely wrong. May be we can reset the counters to zero in the case of failed/killed task attempts. Thoughts?

Reply:
> Do you mean handling the case where the task that got killed has a speculative attempt but got killed because of the speculative attempt succeeded?

Yes, but I understand it's difficult to know precisely.
I don't think that resetting the counter is a good idea as it can confuse users.
My main concern here is that users are confused by the wrongly incremented counter. So one possible compromise is to have note that the counter can be wrongly incremented.

I noticed that `if` block here can also wrongly increment the counter because a speculative task can be zombie.

Reply:
@sarutak I have seen some other counters being reset to zero in the cases where stage attempt is killed or failed. Counters can be wrong only for failed/killed attempts right? Successful attempts should have the counts all correct right except for the one which gets killed with `stage finished` reason but that also only if it is not a speculative task. Isn't it?

Reply:
@sarutak Did you get a chance to think about the above issue? Any thoughts? Gentle reminder

Reply:
@sarutak Gentle reminder. This has been pending for 2 months now. It would be great if you can help me here. Thanks.

Reply:
I re-considered around this logic, and as I commented [here](https://github.com/apache/spark/pull/33253/files#diff-0d63c25e93eb1fbec25754f0bee7714c7e3c1538bc7370d94b147524db5c1f63R762), I'd like to suggest that `killedTaskSummaryForSpeculationStageSummary` should be only for `isSpeculative == true`, and simply have the messages `another attempt succeeded` and `Stage finished` rather than `original attempt succeeded` and `speculated attempt succeeded` ?

Then, we don't need `killedTaskSummaryForSpeculationStageSummary` and can use the existing `killedTaskSummary`.

Reply:
That makes sense too. Even in our internal we don't really use these counters much. I can remove this in that case.